src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1916 @@
+            _options = currentOptions;
+            return expression;
+        }
+
+        private TextSpan GetTokenSpanIncludingEOF(RegexToken token)
+            => token.Kind == RegexKind.EndOfFile
+                ? GetTokenStartPositionSpan(token)
+                : token.GetSpan();
+
+        private TextSpan GetTokenStartPositionSpan(RegexToken token)
+        {
+            return token.Kind == RegexKind.EndOfFile
+                ? new TextSpan(_lexer.Text.Last().Span.End, 0)
+                : new TextSpan(token.VirtualChars[0].Span.Start, 0);
+        }
+
+        private RegexGroupingNode ParseGroupQuestion(RegexToken openParenToken, RegexToken questionToken)
+        {
+            var optionsToken = _lexer.TryScanOptions();
+            if (optionsToken != null)
+            {
+                return ParseOptionsGroupingNode(openParenToken, questionToken, optionsToken.Value);
+            }
+
+            var afterQuestionPos = _lexer.Position;
+
+            // Lots of possible options when we see (?.  Look at the immediately following character
+            // (without any allowed spaces) to decide what to parse out next.
+            ConsumeCurrentToken(allowTrivia: false);
+            switch (_currentToken.Kind)
+            {
+                case RegexKind.LessThanToken:
+                    // (?<=...) or (?<!...) or (?<...>...) or (?<...-...>...)
+                    return ParseLookbehindOrNamedCaptureOrBalancingGrouping(openParenToken, questionToken);
+
+                case RegexKind.SingleQuoteToken:
+                    //  (?'...'...) or (?'...-...'...)
+                    return ParseNamedCaptureOrBalancingGrouping(
+                        openParenToken, questionToken, _currentToken);
+
+                case RegexKind.OpenParenToken:
+                    // alternation construct (?(...) | )
+                    return ParseConditionalGrouping(openParenToken, questionToken);
+
+                case RegexKind.ColonToken:
+                    return ParseNonCapturingGroupingNode(openParenToken, questionToken);
+
+                case RegexKind.EqualsToken:
+                    return ParsePositiveLookaheadGrouping(openParenToken, questionToken);
+
+                case RegexKind.ExclamationToken:
+                    return ParseNegativeLookaheadGrouping(openParenToken, questionToken);
+
+                case RegexKind.GreaterThanToken:
+                    return ParseNonBacktrackingGrouping(openParenToken, questionToken);
+
+                default:
+                    if (_currentToken.Kind != RegexKind.CloseParenToken)
+                    {
+                        // Native parser reports "Unrecognized grouping construct", *except* for (?)
+                        openParenToken = openParenToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                            WorkspacesResources.Unrecognized_grouping_construct,
+                            openParenToken.GetSpan()));
```

Comment:
Should the diagnostic span contain the `_currentToken` as well?

Reply:
i'm fine with span just being on hte ```(```.  it's basically saying "this is some sort of group i do not understand."

Reply:
#won'tfix